src/api.ts

```diff
@@ -0,0 +1,410 @@
+import * as _ from "lodash";
+import * as querystring from "querystring";
+import * as requestModule from "request";
+import * as url from "url";
+import * as FormData from "form-data";
+
+import { Constants } from "./emulator/constants";
+import { FirebaseError } from "./error";
+import * as logger from "./logger";
+import * as responseToError from "./responseToError";
+import * as scopes from "./scopes";
+import * as utils from "./utils";
+
+export enum HttpMethod {
+  GET = "GET",
+  PUT = "PUT",
+  POST = "POST",
+  DELETE = "DELETE",
+  PATCH = "PATCH",
+}
+
+interface UploadFile extends FormData.AppendOptions {
+  stream: unknown;
+}
+
+interface FirebaseRequestLogOptions {
+  skipQueryParams?: boolean;
+  skipRequestBody?: boolean;
+  skipResponseBody?: boolean;
+}
+
+interface FirebaseRequestOptions {
+  auth?: boolean;
+  data?: any; // eslint-disable-line @typescript-eslint/no-explicit-any
+  files?: { [k: string]: UploadFile };
+  form?: { [k: string]: any }; // eslint-disable-line @typescript-eslint/no-explicit-any
+  headers?: requestModule.Headers;
+  json?: any; // eslint-disable-line @typescript-eslint/no-explicit-any
```

Comment:
Shouldn't this be a boolean?

Reply:
`json` in `request` can be an object that is json-encoded and sent in a POST. that's _not_ the case for [`got`](https://www.npmjs.com/package/got) (where I intend to go later). I may be able to consolidate usages of `json` around the codebase, but it's... tedious